lang/java/reef-common/src/main/java/org/apache/reef/runtime/common/driver/evaluator/Evaluators.java

```diff
@@ -118,11 +127,35 @@ public synchronized void put(
    */
   public synchronized void put(final EvaluatorManager evaluatorManager) {
     final String evaluatorId = evaluatorManager.getId();
+    if (this.closedEvaluators.contains(evaluatorId)) {
+      throw new IllegalArgumentException(
+        "Trying to re-add an Evaluator that has already been closed: " + evaluatorId);
+    }
     final EvaluatorManager prev = this.evaluators.put(evaluatorId, evaluatorManager);
     LOG.log(Level.FINEST, "Adding: {0} previous: {1}", new Object[]{evaluatorId, prev});
     if (prev != null) {
       throw new IllegalArgumentException(
-          "Trying to re-add an Evaluator that is already known: " + evaluatorId);
+        "Trying to re-add an Evaluator that is already known: " + evaluatorId);
+    }
+  }
+
+  /**
+   * Moves evaluator from map of active evaluators to set of closed evaluators.
+   */
+  public synchronized void removeClosedEvaluator(final EvaluatorManager evaluatorManager) {
+    final String evaluatorId = evaluatorManager.getId();
+    if (!evaluatorManager.isClosed()) {
+      throw new IllegalArgumentException("Trying to remove evaluator " + evaluatorId + " which is not closed yet.");
+    }
+    if (!this.evaluators.containsKey(evaluatorId) && !this.closedEvaluators.contains(evaluatorId)) {
+      throw new IllegalArgumentException("Trying to remove unknown evaluator " + evaluatorId + ".");
+    }
+    if (!this.evaluators.containsKey(evaluatorId)) {
+      LOG.log(Level.FINE, "Trying to remove closed evaluator " + evaluatorId + " which has already been removed.");
```

Comment:
On line 150 we check that if it's neither in `evaluators` nor in `closedEvaluators` we throw exception. So by now if it's not in `evaluators`, it must be in `closedEvaluators` (being in synchronized method).
I'll add condition to improve readability